lib/node-token-validation/src/api/TokenValidator.ts

```diff
@@ -0,0 +1,138 @@
+        // Prioritize keystore or jwksUri if provided
+        if (validationParams.issuerSigningKeys) {
+            this.logger.verbose("JweKeyStore provided");
+            return validationParams.issuerSigningKeys;
+        } 
+        
+        if (validationParams.issuerSigningJwksUri) {
+            this.logger.verbose("JwksUri provided");
+            return createRemoteJWKSet(new URL(validationParams.issuerSigningJwksUri));
+        }
+
+        // Do resiliency well-known endpoint thing here
+        const retrievedJwksUri: string = await this.openIdConfigProvider.fetchJwksUriFromEndpoint();
+        this.logger.verbose("Creating Jwks from default");
+        return createRemoteJWKSet(new URL(retrievedJwksUri));
+    
+    }
+ 
+    async validateClaims(payload: JWTPayload, validationParams: ValidationParameters): Promise<void> {
+        this.logger.verbose("validateClaims called");
+
+        // Validate nonce
+        if (payload.nonce) {
+            if (!validationParams.nonce) {
+                this.logger.verbose("Nonce detected on token, but nonce not set in validationParams");
+            } else if (validationParams.nonce === payload.nonce) {
+                this.logger.verbose("Nonce validated");
+            } else {
+                throw ValidationError.createInvalidNonceError();
+            }
+        }
+
+        // Validate c_hash
+        if (payload.c_hash && typeof payload.c_hash === "string") {
+            this.logger.verbose("Validating c_hash");
+
+            if (!validationParams.code) {
+                this.logger.verbose("C_hash present on token but code not set in validationParams. Unable to validate c_hash");
+            } else {
+                const hashResult = await this.checkHashValue(validationParams.code, payload.c_hash);
+                if (!hashResult) {
+                    throw ValidationError.createInvalidCHashError();
+                }
+            }
+        }
+
+        // Validate at_hash
+        if (payload.at_hash && typeof payload.at_hash === "string") {
+            this.logger.verbose("Validating at_hash");
+
+            if (!validationParams.accessTokenForAtHash) {
+                this.logger.verbose("At_hash present on token but access token not set in validationParams. Unable to validate at_hash");
+            } else {
+                const hashResult = await this.checkHashValue(validationParams.accessTokenForAtHash, payload.at_hash);
+                if (!hashResult) {
+                    throw ValidationError.createInvalidAtHashError();
+                }
+            }
+        }
+    }
+
+    async checkHashValue(content: string, hashProvided: string): Promise<Boolean> {
+        this.logger.verbose("checkHashValue called");
+
+        // 1. Hash the content (either code for c_hash, or token for at_hash) and save as buffer
+        const digest = crypto.createHash("sha256").update(content, "ascii").digest();
+
+        // 2. Only take first half of buffer
```

Comment:
Why only half? (would be good to put answer to that in the comment as well)

Appreciate teh comments!

Reply:
This is how the OIDC spec specifies `c_hash` and `at_hash` should be validated. See [this link](https://openid.net/specs/openid-connect-core-1_0.html#HybridTokenValidation) at 3.3.2.11.

Added this link to the comments for clarity!